packages/search-in-workspace/src/common/search-in-workspace-service.ts

```diff
@@ -0,0 +1,125 @@
+                @inject(SearchInWorkspaceClientImpl) protected readonly client: SearchInWorkspaceClientImpl,
+                @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService,
+                @inject(ILogger) protected readonly logger: ILogger,
+        ) {
+                client.setService(this);
+        }
+
+        onResult(searchId: number, result: SearchInWorkspaceResult): void {
+                const callbacks = this.pendingSearches.get(searchId);
+
+                if (callbacks) {
+                        callbacks.onResult(searchId, result);
+                }
+        }
+
+        onDone(searchId: number, error?: string): void {
+                const callbacks = this.pendingSearches.get(searchId);
+
+                if (callbacks) {
+                        this.pendingSearches.delete(searchId);
+                        callbacks.onDone(searchId, error);
+                } else {
+                        if (searchId > this.lastKnownSearchId) {
+                                this.logger.debug(`Got an onDone for a searchId we don't know about (${searchId}), stashing it for later with error = `, error);
+                                this.pendingOnDones.set(searchId, error);
+                        } else {
+                                /* It's possible to receive an onDone for a search we have cancelled.  Just ignore it.  */
+                                this.logger.debug(`Got an onDone for a searchId we don't know about (${searchId}), but it's probably an old one, error = `, error);
+                        }
+                }
+        }
+
+        /* Start a search of the string "what" in the workspace.  */
+        async search(what: string, callbacks: SearchInWorkspaceCallbacks, opts?: SearchInWorkspaceOptions): Promise<number> {
+                const root = await this.workspaceService.root;
+
+                if (!root) {
+                        throw new Error("Search failed: no workspace root.");
+                }
+
+                const rootUri = new URI(root.uri);
+                const searchId = await this.searchServer.search(what, rootUri.path.toString(), opts);
+                this.pendingSearches.set(searchId, callbacks);
+                this.lastKnownSearchId = searchId;
+
+                this.logger.debug('Service launched search ' + searchId);
+
+                /* Check if we received an onDone before search() returned.  */
+                if (this.pendingOnDones.has(searchId)) {
+                        this.logger.debug('Ohh, we have a stashed onDone for that searchId');
+                        const error = this.pendingOnDones.get(searchId);
+                        this.pendingOnDones.delete(searchId);
+
+                        /* Call the client's searchId, but first give it a
+                           chance to record the returned searchId.  */
+                        setTimeout(() => {
```

Comment:
use process.nextTick ?

Reply:
This runs in the browser, so it's not available.

Reply:
Right